### PR TITLE
add script to check duplicates entry in path

### PR DESCRIPTION
**Pull request by k-okada**
as suggested at https://github.com/ros-infrastructure/rosdep/pull/520#pullrequestreview-44476894
added argumetns to allow take local files

```
./scripts/check_duplicates.py rosdep/*.yaml
```

I also added this to test, but I'm afraid this is too aggresive

@tfoote

### CODE REVIEW

**Comment by mikaelarguedas** [comment]
please rebase on top on current master, to provide a PR with a single commit https://github.com/ros/rosdistro/pull/15263/commits/67963d8a6b0dd60de19dbadc377e786f03bee7fa

Please see the travis log to see the reason of the failure

**Comment by k-okada** [comment]
fixed and passed test, but that's strange, because the test should be failed.
may be we do need not run `rosdep init  & rosdep update` to enable this test code.

**Comment by k-okada** [comment]
ok, fixed, now it fails as expected, need to merge #15260 #15259  to pass this test.
```

checking with ['indigo', 'ubuntu', 'trusty']
checking sources
- file:///home/travis/build/ros/rosdistro/indigo/distribution.yaml
- flie://home/travis/build/ros/rosdistro/rosdep/base.yaml
- flie://home/travis/build/ros/rosdistro/rosdep/gentoo.yaml
- flie://home/travis/build/ros/rosdistro/rosdep/python.yaml
- flie://home/travis/build/ros/rosdistro/rosdep/ruby.yaml
checking duplicates
* file:///home/travis/build/ros/rosdistro/indigo/distribution.yaml
* flie://home/travis/build/ros/rosdistro/rosdep/base.yaml
  ERR: pocketsphinx is multiply defined in
	file:///home/travis/build/ros/rosdistro/indigo/distribution.yaml and 
	flie://home/travis/build/ros/rosdistro/rosdep/base.yaml
* flie://home/travis/build/ros/rosdistro/rosdep/gentoo.yaml
* flie://home/travis/build/ros/rosdistro/rosdep/python.yaml
  ERR: xdot is multiply defined in
	file:///home/travis/build/ros/rosdistro/indigo/distribution.yaml and 
	flie://home/travis/build/ros/rosdistro/rosdep/python.yaml
```

**Comment by k-okada** [comment]
@mikaelarguedas @tfoote ik, now we passed the travis test so ready fo a reivew, 
I confirmed this PR failes if we revert commit 2993ff6105b8efe59517936d7f8469ce44307335
```
* flie://home/travis/build/ros/rosdistro/rosdep/ruby.yaml
* flie://home/travis/build/ros/rosdistro/rosdep/python.yaml
  ERR: xdot is multiply defined in
	file:///home/travis/build/ros/rosdistro/indigo/distribution.yaml and 
	flie://home/travis/build/ros/rosdistro/rosdep/python.yaml
* sources.list
checking with ['jade', 'ubuntu', 'trusty']
checking sources
```
https://travis-ci.org/ros/rosdistro/builds/251355160

I have to install python-rosdep to run this feature, but I am not confident in this implementation, if someone   know better way, let me know.

File: scripts/check_duplicates.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    else:
+        print(text)
+
+
+def print_test(msg):
+    printc(msg, 'yellow')
+
+
+def print_err(msg):
+    printc('  ERR: ' + msg, 'red')
```
Comment (dirk-thomas) [inline scripts/check_duplicates.py]: I would rather avoid adding this colorization logic just for this test. None of the other tests does it either.

File: scripts/check_duplicates.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    ret = True
+    for tag in [['indigo', 'ubuntu', 'trusty'],
+                ['jade', 'ubuntu', 'trusty'],
+                ['kinetic', 'ubuntu', 'xenial'],
+                ['lunar', 'ubuntu', 'xenial']]:
+        matcher.tags = tag
+        print_test('checking with %s'%(matcher.tags))
+        sources = [x for x in sources if matcher.matches(x)]
+        ret = ret & check_duplicates(sources)
+    return ret
```
Comment (dirk-thomas) [inline scripts/check_duplicates.py]: Please return an integer from the `main` function which represents an error code or `0` if successful.

File: scripts/check_duplicates.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Checks whether rosdep files contains duplicate ROS rules')
+    parser.add_argument('infiles', nargs='*', help='input rosdep YAML file')
+    args = parser.parse_args()
+    if not main(args.infiles):
+        sys.exit(1)
```
Comment (dirk-thomas) [inline scripts/check_duplicates.py]: This can just be `sys.exit(main(...))` then.

File: test/rosdep_duplicates_test.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import os
+
+from scripts.check_duplicates import main as check_duplicates
+
+from .fold_block import Fold
+
+
+def test():
+    files = os.listdir('rosdep')
+    files = filter(lambda x: x.endswith('.yaml'), files) # accept only file ends with .yaml
```
Comment (dirk-thomas) [inline test/rosdep_duplicates_test.py]: Please use a list comprehension for this instead. Also please follow PEP 8 for the comments (two spaces + `#` + one space + comment text)

File: test/rosdep_duplicates_test.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    files = [os.path.join('rosdep', x) for x in files] # use relative path
+    with Fold() as fold:
+        print("""Running 'scripts/check_duplicates.py' on all '*.yaml' in the 'rosdep' directory.
```
Comment (dirk-thomas) [inline test/rosdep_duplicates_test.py]: Please use single `"` here. No need to use `"""`.

Comment (k-okada) [inline test/rosdep_duplicates_test.py]: @dirk-thomas thank you for comment, I think I have fixed all parts

File: scripts/check_duplicates.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        print("- %s" % source.url)
+
+    # replace with infile
+    for filename in infile:
+        filepath = os.path.join(os.getcwd(), filename)
+        with open(filepath) as f:
+            rosdep_data = yaml.load(f.read())
+        # osx-homebrow uses xos tag
+        tag = 'osx' if 'osx-' in filepath else ''
+        model = CachedDataSource('yaml', 'flie:/'+filepath, [tag], rosdep_data)
```
Comment (dirk-thomas) [inline scripts/check_duplicates.py]: Spelling: `flie`

**Review by dirk-thomas — COMMENTED** [review]
CI is currently failing.

File: scripts/check_duplicates.py (hunk 2 above)
Comment (k-okada) [inline scripts/check_duplicates.py]: revert this commit https://github.com/ros/rosdistro/pull/15263/commits/00fd68889ac1b5d7779a68aab589ab91cd0cf3bf
It seems nosetest need to return True or False, not integer,

Comment (dirk-thomas) [inline scripts/check_duplicates.py]: I am not sure I can follow: how is the `main` function being called from `nosetests`? `nosetests` should only pick up the test functions defined in `rosdep_duplicates_test.py`.

Comment (k-okada) [inline scripts/check_duplicates.py]: @dirk-thomas yes and I called `main` as `check_duplicates`at the last of `rosdep_duplicates_test.py`
https://github.com/ros/rosdistro/pull/15263/files#diff-857189dd46c3a32ffe6a1424eaf41c5fR5

I just followed https://github.com/ros/rosdistro/blob/master/test/rosdep_formatting_test.py and https://github.com/ros/rosdistro/blob/master/scripts/check_rosdep.py style, so please let me know your recommended test and script code, I can catch your direction.

**Review by dirk-thomas — APPROVED** [review]

**Comment by k-okada** [comment]
@dirk-thomas thank you!!

**Comment by dirk-thomas** [comment]
Thank you for your effort coming up with the patch!

**Pull request merged**